File: 11.05-arrays/matrix.py
import sys
import random


def randomize(size):
    """
    Create a square matrix of random one-digit integers.

    :param size: A desired dimension of the matrix
    :param return: A random matrix of that dimension
    """
    return [[random.randint(0, 9) for _ in range(size)] for _ in range(size)]


def multiply(matrix_a, matrix_b):
    """
    Mupliply two matrices.

    :param matrix_a: A first square matrix
    :param matrix_b: A second matrix of the same dimensions
    :return: The product of the two matrices
    """
    size = len(matrix_a)
    matrix_c = [[0 for _ in range(size)] for _ in range(size)]

    for i in range(size):
        for j in range(size):
            for k in range(size):
                matrix_c[i][j] += matrix_a[i][k] * matrix_b[k][j]

    return matrix_c


def main():
    size = int(sys.argv[1])
    matrix_a = randomize(size)
    matrix_b = randomize(size)
    matrix_c = multiply(matrix_a, matrix_b)


if __name__ == "__main__":
    main()

File: 11.05-arrays/matrix.c
#include <stdlib.h>
#include <stdio.h>

int *mcreate(int);
int *mmult(int *, int *, int);

int main(int argc, char *argv[]) {
    int n, *mat_a, *mat_b, *mat_c;

    n = strtol(argv[1], NULL, 10);
    mat_a = mcreate(n);
    mat_b = mcreate(n);
    mat_c = mmult(mat_a, mat_b, n);

    free(mat_a);
    free(mat_b);
    free(mat_c);

    return 0;
}

/* mcreate: Creates an n x n matrix of random one-digit integers. */
int *mcreate(int n) {
    int i, j, *mat;

    mat = (int *)calloc(sizeof(int), n * n);

    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            mat[i * n + j] = rand() % 10;
        }
    }

    return mat;
}

/* mmult: Multiplies two n x n matrices. */
int *mmult(int *mat_a, int *mat_b, int n) {
    int i, j, k, *mat_c;

    mat_c = (int *)calloc(sizeof(int), n * n);

    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            for (k = 0; k < n; k++) {
                mat_c[i * n + j] += mat_a[i * n + k] * mat_b[k * n + j];
            }
        }
    }

    return mat_c;
}
